### PR TITLE
o/snapstate: keep old revision if install doesn't add new one

### PR DESCRIPTION
**Pull request by MiguelPires**
The GC code was assuming that an install always added a new revision
so it accounted for that when removing the oldest revs. Since we may
install a revision that's already stored, this would result in us
removing a revision even though we were under the configured limit.

### CODE REVIEW

**Review by mardy — APPROVED** [review]
LGTM!

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -451,7 +451,9 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 				retain = 3
 			}
 		}
-		retain-- //  we're adding one
+		if snapst.LastIndex(snapsup.Revision()) == -1 {
+			retain-- //  we're adding one
+		}
```
Comment (stolowski) [inline overlord/snapstate/snapstate.go:456]: Could you add a comment here explaining why we need this extra check (basically, what you explained in the PR description could be reused here).

File: overlord/snapstate/snapstate_install_test.go [hunk 2]
```diff
@@ -4246,3 +4246,71 @@ func (s *snapmgrTestSuite) TestInstallDeduplicatesSnapNames(c *C) {
...
+
+	seq := []*snap.SideInfo{{
+		RealName: "some-snap",
+		SnapID:   "some-snap-id",
+		Revision: snap.R(1),
+	}, {
+		RealName: "some-snap",
+		SnapID:   "some-snap-id",
+		Revision: snap.R(2),
+	}}
```
Comment (stolowski) [inline overlord/snapstate/snapstate_install_test.go:4282]: I think we do have a test that checks installation of **different** revisions and retaining the correct number of them right, no need to test that?

**Review by stolowski — COMMENTED** [review]
Looks nice, thanks for investigation this. Two minor comments.

File: overlord/snapstate/snapstate.go (hunk 1 above)
Comment (pedronis) [reply overlord/snapstate/snapstate.go:456]: snapsup.Revision() is copied to  targetRevision earlier, also there's a loop below that seems to be checking a similar thing to the new "if" here, so perhaps this should be integrated there somehow?

```
		// make sure we're not scheduling the removal of the target
		// revision in the case where the target revision is already in
		// the sequence.
		for i := 0; i < currentIndex; i++ {
```

**Review by pedronis — COMMENTED** [review]
question

File: overlord/snapstate/snapstate_install_test.go (hunk 2 above)
Comment (MiguelPires) [reply overlord/snapstate/snapstate_install_test.go:4282]: True, this can be simplified to just one revision. Thanks

Comment (stolowski) [reply overlord/snapstate/snapstate_install_test.go]: Hmm, that's fine but it wasn't my question :)

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -451,6 +451,8 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 				retain = 3
 			}
 		}
+
+		// if we're not using a stored revision, account for the one being added
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:455]: maybe s/store/already present/ ?

**Review by pedronis — APPROVED** [review]
thanks, small consideration

File: overlord/snapstate/snapstate_install_test.go (hunk 2 above)
Comment (MiguelPires) [reply overlord/snapstate/snapstate_install_test.go]: Oh sorry, I misunderstood then. I thought your point was that the test doesn't need more than one revision, it just needs to test that installing the same revision doesn't the discard the existing one

Comment (stolowski) [reply overlord/snapstate/snapstate_install_test.go]: Yes, sorry, indeed my question wasn't clear. I was asking whether we need to check for regular cases of garbage-collection of old revisions. But I just checked and we have a bunch od update tests about that already, so it's fine.

**Review by stolowski — APPROVED** [review]
Thank you! LGTM

**Pull request merged**